File: backend/app/api/endpoints/clients.py
"""
Client endpoints module.

This module contains endpoints for client profile management.
"""
from typing import Any, List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session

from app import crud
from app.api.deps import get_current_admin_user, get_current_user, get_session
from app.models.client import ClientProfile, ClientProfileCreate, ClientProfileRead, ClientProfileUpdate
from app.models.user import User, UserRole

router = APIRouter()


@router.get("/", response_model=List[ClientProfileRead])
def read_clients(
    *,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_admin_user),
    skip: int = 0,
    limit: int = 100,
    active_only: bool = True,
) -> Any:
    """
    Retrieve all client profiles.
    
    Only accessible by admin users.
    """
    clients = crud.get_clients(
        session, skip=skip, limit=limit, active_only=active_only
    )
    return clients


@router.post("/", response_model=ClientProfileRead)
def create_client(
    *,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_admin_user), # Ensure only admin can create
    client_in: ClientProfileCreate, # Use the updated schema with user details
) -> Any:
    """
    Create a new client user and their associated profile.
    
    Only accessible by admin users.
    """
    # The CRUD function now handles user creation and email checks
    try:
        client = crud.create_client(session=session, obj_in=client_in)
        # Eager load user data for the response model if necessary
        # Accessing client.user might trigger a lazy load if not handled in CRUD
        # or if the session is closed before serialization.
        # Let's assume ClientProfileRead doesn't need deep user details beyond user_id
        return client
    except HTTPException as e:
        # Re-raise HTTPExceptions (like email exists) from CRUD
        raise e
    except ValueError as e: # Catch potential ValueErrors from CRUD if any remain
         raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        # Catch unexpected errors
        print(f"Unexpected error creating client: {e}") # Log for debugging
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Internal server error")


@router.get("/{client_id}", response_model=ClientProfileRead)
def read_client(
    *,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
    client_id: int,
) -> Any:
    """
    Get a specific client profile by ID.
    
    Accessible by:
    - Admin users (can access any client profile)
    - Client users (can only access their own client profile)
    """
    client = crud.get_client(session, client_id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client profile not found",
        )
    
    # Check permissions
    if current_user.role != UserRole.ADMIN and client.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions",
        )
    
    return client


@router.put("/{client_id}", response_model=ClientProfileRead)
def update_client(
    *,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_admin_user),
    client_id: int,
    client_in: ClientProfileUpdate,
) -> Any:
    """
    Update a client profile.
    
    Only accessible by admin users.
    """
    client = crud.get_client(session, client_id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client profile not found",
        )
    
    client = crud.update_client(session, db_obj=client, obj_in=client_in)
    return client


@router.delete("/{client_id}", response_model=ClientProfileRead)
def delete_client(
    *,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_admin_user),
    client_id: int,
) -> Any:
    """
    Delete a client profile.
    
    Only accessible by admin users.
    """
    client = crud.get_client(session, client_id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client profile not found",
        )
    
    client = crud.delete_client(session, client_id=client_id)
    return client


@router.get("/me/", response_model=ClientProfileRead)
def read_client_me(
    *,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Get the client profile of the current user.
    
    Only accessible by client users.
    """
    # Check if user is a client
    if current_user.role != UserRole.CLIENT:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not a client user",
        )
    
    # Get client profile
    client = crud.get_by_user_id(session, current_user.id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client profile not found",
        )
    
    return client

File: frontend/src/services/clients.ts
import api from './api';

export interface ClientProfile {
  id: number;
  user_id: number;
  company_name: string;
  industry: string;
  website?: string;
  linkedin_url?: string;
  description?: string;
  logo_url?: string;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

// Interface matching the Pydantic schema + user details for creation request body
export interface ClientProfileCreate {
  // user_id is removed, user details are added
  email: string; 
  password: string;
  full_name?: string;
  company_name: string;
  industry: string;
  website?: string;
  linkedin_url?: string;
  description?: string;
  logo_url?: string;
  is_active?: boolean;
}

export interface ClientProfileUpdate {
  company_name?: string;
  industry?: string;
  website?: string;
  linkedin_url?: string;
  description?: string;
  logo_url?: string;
  is_active?: boolean;
}

const clientService = {
  /**
   * Get all client profiles
   * @returns Promise with list of client profiles
   */
  getAll: async (): Promise<ClientProfile[]> => {
    const response = await api.get<ClientProfile[]>('/api/v1/clients/');
    return response.data;
  },

  /**
   * Get a client profile by ID
   * @param id Client profile ID
   * @returns Promise with client profile
   */
  getById: async (id: number): Promise<ClientProfile> => {
    const response = await api.get<ClientProfile>(`/api/v1/clients/${id}`);
    return response.data;
  },

  /**
   * Get the current user's client profile
   * @returns Promise with client profile
   */
  getMyProfile: async (): Promise<ClientProfile> => {
    const response = await api.get<ClientProfile>('/api/v1/clients/me/');
    return response.data;
  },

  /**
   * Create a new client profile
   * @param data Client profile data
   * @returns Promise with created client profile
   */
  create: async (data: ClientProfileCreate): Promise<ClientProfile> => {
    const response = await api.post<ClientProfile>('/api/v1/clients/', data);
    return response.data;
  },

  /**
   * Update a client profile
   * @param id Client profile ID
   * @param data Client profile update data
   * @returns Promise with updated client profile
   */
  update: async (id: number, data: ClientProfileUpdate): Promise<ClientProfile> => {
    const response = await api.put<ClientProfile>(`/api/v1/clients/${id}`, data);
    return response.data;
  },

  /**
   * Delete a client profile
   * @param id Client profile ID
   * @returns Promise with deleted client profile
   */
  delete: async (id: number): Promise<ClientProfile> => {
    const response = await api.delete<ClientProfile>(`/api/v1/clients/${id}`);
    return response.data;
  }
};

export default clientService;
